### ISSUE TITLE
internal error when adding param string to string literal in 'external-name' for extern statement with "always resolve function" pragma

### ISSUE THREAD

**Issue by stonea** [chapel-lang/chapel#22224]
### Summary of Problem

The following Chapel code triggers an internal error:

```chpl
proc foo(param opName : string) {
  pragma "always resolve function"
  extern "prefix" + opName proc someExternFunc();
}

foo("suffix");
```

```
CHPL_HOME/modules/internal/String.chpl:2143: In method '+':
$CHPL_HOME/modules/internal/String.chpl:2144: internal error: AST-EXP-1144 chpl version 1.31.0 pre-release (90636e61f7)

Internal errors indicate a bug in the Chapel compiler,
and we're sorry for the hassle.  We would appreciate your reporting this bug --
please see https://chapel-lang.org/bugs.html for instructions.  In the meantime,
the filename + line number above may be useful in working around the issue.

  foo.chpl:3: called as +(param a = "prefix", param b = )
note: generic instantiations are underlined in the above callstack
```

Interestingly enough the issues goes away if I remove the `always resolve function` pragma.

### Configuration Information

<details>

```
$ chpl --version
warning: The prototype GPU support implies --no-checks. This may impact debuggability. To suppress this warning, compile with --no-checks explicitly
chpl version 1.31.0 pre-release (6f269987fc)
  built with LLVM version 14.0.6
  available LLVM targets: xcore, x86-64, x86, wasm64, wasm32, ve, systemz, sparcel, sparcv9, sparc, riscv64, riscv32, ppc64le, ppc64, ppc32le, ppc32, nvptx64, nvptx, msp430, mips64el, mips64, mipsel, mips, lanai, hexagon, bpfeb, bpfel, bpf, avr, thumbeb, thumb, armeb, arm, amdgcn, r600, aarch64_32, aarch64_be, aarch64, arm64_32, arm64
Copyright 2020-2023 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)

$ $CHPL_HOME/util/printchplenv --anonymize
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: llvm *
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: gpu *
CHPL_COMM: none *
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: system *
CHPL_AUX_FILESYS: none
```

</details>

**Comment by stonea** [comment]
To give a little more stack information here's what I get with `CHPL_DEVELOPER=1`

```
$CHPL_HOME/modules/internal/String.chpl:2143: In method '+':
$CHPL_HOME/modules/internal/String.chpl:2144: internal error: string literal expression expected [AST/expr.cpp:1144]
  foo.chpl:3: called as +(param a = "prefix", param b = )
note: generic instantiations are underlined in the above callstack
```

**Comment by e-kayrakli** [comment]
Maybe the intention is not to support that? Because putting the string in a `param` first gives a slightly more meaningful error:

```chapel
proc foo(param opName : string) {
  param externRename = "prefix" + opName;
  pragma "always resolve function"
  extern externRename proc someExternFunc();
}

foo("suffix");
```

```
error: extern name expression must be a string
```

Which I interpret as "string literal" :)

**Comment by stonea** [comment]
> Maybe the intention is not to support that? Because putting the string in a param first gives a slightly more meaningful error:

Maybe, it's a bit surprising what works and what doesn't. It seems to be more complicated than "no non literal strings allowed for the 'external-name' expression" though.  I imagine it's a matter of what gets folded to a literal string before we process the extern statement, but I don't really have a clear sense of what the ordering of this processing is.

For example this works:

```chpl
extern {
  static void prefix_a() { printf("Hello world A\n"); }
  static void prefix_b() { printf("Hello world B\n"); }
}

config param N = 0;
pragma "always resolve function"
extern "prefix" + if N < 0 then "_a" else "_b" proc someExternFunc();
someExternFunc();
```

But this doesn't:

```
// ...

config param N = 0;
param externRename = "prefix" + if N < 0 then "_a" else "_b"
pragma "always resolve function"
extern externRename proc someExternFunc();
someExternFunc();
```

And in both cases if I remove the pragma it's fine.

**Comment by e-kayrakli** [comment]
Yeah, I agree it has something to do with the order of resolution, and that pragma is impacting that somehow.

FWIW, even if this was not intended to be supported initially, it looks like a powerful feature.

**Comment by stonea** [comment]
> FWIW, even if this was not intended to be supported initially, it looks like a powerful feature.

What led me to bumping into this is the use of a function in our atomics runtime lib that computes the name for an extern function (see https://github.com/chapel-lang/chapel/blob/36908199a56b60bb55c802b8821fe81767aed533/modules/internal/Atomics.chpl#LL211C16-L211C26 and its uses throughout `Atomics.chpl`). Our [language spec](https://chapel-lang.org/docs/language/spec/interoperability.html?highlight=extern#calling-external-functions) also says "The external-name expression must evaluate to a param string". So I think there is some precedent of 'external-name' being something other than a string literal.

I suppose to lawyer it, I could argue that: "must evaluate to a param string" doesn't strictly mean "will work for all param strings", but it seems like a reasonable assumption to think that if `param x = ...` evaluates fine than `extern x proc someProc()` should work as well.

@bradcray / @lydia-duncan, is stabilizing the behavior of `external-name` in an `extern` statement something we should/need-to be concerned about for Chapel 2.0?

I guess since I'm only encountering this if I use `pragma "always resolve function"` that would mean the behavior I'm seeing would be outside of what we expect users to encounter (the behavior of individual pragmas isn't something included in the definition of Chapel 2.0 right?).

**Comment by lydia-duncan** [comment]
That seems like a bug to me, so I wouldn't view it as reconsidering the behavior of `external name` in an `extern` statement w.r.t. 2.0.

**Comment by lydia-duncan** [comment]
> the behavior of individual pragmas isn't something included in the definition of Chapel 2.0 right?

Correct.  Pragmas are not intended to be user-facing

**Comment by bradcray** [comment]
I believe that only literal strings were intended to be supported here in the original implementation (and that that's what the internal error is trying to say — but in retrospect, it should've been a user-facing error which would arguably change this from a bug to a feature request).  If no cases worked, I'd agree that this is not a 2.0 issue since it wouldn't break existing codes if we enabled it later.  The fact that Andy's able to get some cases through is a bit concerning because if we decided "We only intended to ever support string literals" then codes that work today will cease to.  Though if the documentation says it must be a string literal, then we're covered again because we can say the fact that some cases got through was itself a bug.

**Comment by bradcray** [comment]
Oops, though I may be completely wrong.  Or if I was right at some point, I may be outdated.  Specifically, I'm noting that the spec merely says the extern proc name needs to be a `param string` without requiring it to be a literal and we have working tests that rely on this (notably `extern/ferguson/param-cname[-parens].chpl`).  So I agree that this is a bug rather than more than we intended.
